### DISCUSSION TITLE
Aerie Ecosystem

### DISCUSSION THREAD

**Discussion by camargo** [NASA-AMMOS/aerie#575] [Show and tell]
This document lists all the software components in the Aerie ecosystem.

## Java Libraries

1. [constraints](https://github.com/NASA-AMMOS/aerie/tree/develop/constraints) - Java library for constraint checking
1. [contrib](https://github.com/NASA-AMMOS/aerie/tree/develop/contrib) - Java library of common convenience classes for mission models
1. [merlin-driver](https://github.com/NASA-AMMOS/aerie/tree/develop/merlin-driver) - Java library for discrete-event simulation execution
1. [merlin-framework-junit](https://github.com/NASA-AMMOS/aerie/tree/develop/merlin-framework-junit) - Java library extension of JUnit to unit test mission models
1. [merlin-framework-processor](https://github.com/NASA-AMMOS/aerie/tree/develop/merlin-framework-processor) - Java library for annotation processing in mission models
1. [merlin-framework](https://github.com/NASA-AMMOS/aerie/tree/develop/merlin-framework) - Java library for writing mission models
1. [merlin-sdk](https://github.com/NASA-AMMOS/aerie/tree/develop/merlin-sdk) - Java library interface between mission models and the merlin-driver
1. [parsing-utilities](https://github.com/NASA-AMMOS/aerie/tree/develop/parsing-utilities) - Java library for JSON serialization and deserialization
1. [scheduler-driver](https://github.com/NASA-AMMOS/aerie/tree/develop/scheduler-driver) - Java library for goal-oriented scheduling execution

## JavaScript/TypeScript Libraries

1. [aerie-ampcs](https://github.com/NASA-AMMOS/aerie-ampcs) - Utility functions to convert a standard XML AMPCS command dictionary to JavaScript
1. [aerie-ts-user-code-runner](https://github.com/NASA-AMMOS/aerie-ts-user-code-runner) - Library to safely run user code written in TypeScript

## Services

1. [aerie-gateway](https://github.com/NASA-AMMOS/aerie-gateway) - Service for managing files and authentication
1. [merlin-server](https://github.com/NASA-AMMOS/aerie/tree/develop/merlin-server) - Service for planning and simulation
1. [merlin-worker](https://github.com/NASA-AMMOS/aerie/tree/develop/merlin-worker) - Service that executes concurrent simulation requests
1. [scheduler-server](https://github.com/NASA-AMMOS/aerie/tree/develop/scheduler-server) - Service for scheduling
1. [scheduler-worker](https://github.com/NASA-AMMOS/aerie/tree/develop/scheduler-worker) - Service that executes concurrent scheduling requests
1. [sequencing-server](https://github.com/NASA-AMMOS/aerie/tree/develop/sequencing-server) - Service for sequence generation and management

## Applications and Tools

1. [aerie-cli](https://github.com/NASA-AMMOS/aerie-cli) - CLI tool for Aerie API convenience use
1. [aerie-docs](https://github.com/NASA-AMMOS/aerie-docs) - Web-based documentation website for Aerie
1. [aerie-ui](https://github.com/NASA-AMMOS/aerie-ui) - Web-based client application for Aerie

## Example Mission Models

1. [aerie-mission-model-template](https://github.com/NASA-AMMOS/aerie-mission-model-template) - Template project for starting an Aerie mission model
1. [aerie-lander](https://github.com/NASA-AMMOS/aerie-lander) - Example mission model for a lander mission
1. [examples](https://github.com/NASA-AMMOS/aerie/tree/develop/examples) - Different examples of simple Aerie mission models

## End-to-End Tests

1. [db-tests](https://github.com/NASA-AMMOS/aerie/tree/develop/db-tests) - Aerie repo database tests
1. [e2e-tests](https://github.com/NASA-AMMOS/aerie/tree/develop/e2e-tests) - Aerie repo end-to-end tests
1. [e2e-tests](https://github.com/NASA-AMMOS/aerie-ui/tree/develop/e2e-tests) - Aerie UI repo end-to-end tests

**Comment by ewferg** [comment]
I think this is a useful way to break things up and something we should maybe include in the README? I would say this would replace the directory structure section, which could potentially go in a contributing area, but I don't have any strong opinions there. Another way we could break this down that may be a little simpler for an outside audience would be by high level capability where we would combine some of the libraries/services together. What do you think of this:

### Simulation Engine
1. [merlin-driver](https://github.com/NASA-AMMOS/aerie/tree/develop/merlin-driver) - Java library for discrete-event simulation execution
2. [merlin-server](https://github.com/NASA-AMMOS/aerie/tree/develop/merlin-server) - Service for planning and simulation
3. [merlin-worker](https://github.com/NASA-AMMOS/aerie/tree/develop/merlin-worker) - Service that executes concurrent simulation requests

### Mission Model Framework
4. [merlin-framework-junit](https://github.com/NASA-AMMOS/aerie/tree/develop/merlin-framework-junit) - Java library extension of JUnit to unit test mission models
5. [merlin-framework-processor](https://github.com/NASA-AMMOS/aerie/tree/develop/merlin-framework-processor) - Java library for annotation processing in mission models
6. [merlin-framework](https://github.com/NASA-AMMOS/aerie/tree/develop/merlin-framework) - Java library for writing mission models
7. [merlin-sdk](https://github.com/NASA-AMMOS/aerie/tree/develop/merlin-sdk) - Java library interface between mission models and the merlin-driver
8. [contrib](https://github.com/NASA-AMMOS/aerie/tree/develop/contrib) - Java library of common convenience classes for mission models

### Scheduling
1. [scheduler-driver](https://github.com/NASA-AMMOS/aerie/tree/develop/scheduler-driver) - Java library for goal-oriented scheduling execution
2. [scheduler-server](https://github.com/NASA-AMMOS/aerie/tree/develop/scheduler-server) - Service for scheduling
3. [scheduler-worker](https://github.com/NASA-AMMOS/aerie/tree/develop/scheduler-worker) - Service that executes concurrent scheduling requests

### Constraint Checking
1. [constraints](https://github.com/NASA-AMMOS/aerie/tree/develop/constraints) - Java library for constraint checking

### Sequencing
1. [sequencing-server](https://github.com/NASA-AMMOS/aerie/tree/develop/sequencing-server) - Service for sequence generation and management

### Core Infrastructure
1. [parsing-utilities](https://github.com/NASA-AMMOS/aerie/tree/develop/parsing-utilities) - Java library for JSON serialization and deserialization
4. [aerie-ampcs](https://github.com/NASA-AMMOS/aerie-ampcs) - Utility functions to convert a standard XML AMPCS command dictionary to JavaScript
5. [aerie-ts-user-code-runner](https://github.com/NASA-AMMOS/aerie-ts-user-code-runner) - Library to safely run user code written in TypeScript
6. [aerie-gateway](https://github.com/NASA-AMMOS/aerie-gateway) - Service for managing files and authentication

**Reply by camargo** [reply]
Yeah I think it's a good idea to add this list to our doc site. I don't think we should replace the directory structure in the Aerie repo README just because this list includes things outside the directory structure of that repo (and the directory structure also includes things not in this list).
I like the breakdown by higher-level capability. Is there a reason you excluded the application & tools, examples models, and test components?

**Reply by Twisol** [reply]
These two ways of organization remind me of [organizing by feature vs. layer](http://www.javapractices.com/topic/TopicAction.do?Id=205). I tend to like organization by feature better as a default -- although knowing how everything gets composited in layers is useful too :)
